core/src/main/scala/filodb.core/memstore/TimeSeriesShard.scala

```diff
@@ -456,6 +462,54 @@ class TimeSeriesShard(val dataset: Dataset,
 
   def indexValues(indexName: String, topK: Int): Seq[TermInfo] = partKeyIndex.indexValues(indexName, topK)
 
+  /**
+    * This method is to apply column filters and fetch matching label values.
+    */
+  def indexValuesWithFilters(filter: Seq[ColumnFilter],
+                             column: String,
+                             endTime: Long,
+                             startTime: Long,
+                             limit: Int): List[ZeroCopyUTF8String] = {
+    val result: Set[ZeroCopyUTF8String] = new mutable.HashSet[ZeroCopyUTF8String]()
+    val partIterator = partKeyIndex.partIdsFromFilters(filter, startTime, endTime)
+    while(partIterator.hasNext && result.size < limit) {
+      val nextPart = getPartitionFromPartId(partIterator.next())
+      val consumer = new SeqIndexValueConsumer(column)
+      // FIXME This is non-performant and temporary fix for fetching label values based on filter criteria.
+      // Other strategies needs to be evaluated for making this performant - create facets for predefined fields or
+      // have a centralized service/store for serving metadata
+      dataset.partKeySchema.consumeMapItems(nextPart.partKeyBase, nextPart.partKeyOffset, 0, consumer)
+      result ++= consumer.labelValues
+    }
+    result.toList
+  }
+
+  /**
+    * This method is to apply column filters and fetch matching time series partition keys.
+    */
+  def partKeysWithFilters(filter: Seq[ColumnFilter],
+                             endTime: Long,
+                             startTime: Long,
+                             limit: Int): List[ZeroCopyUTF8String] = {
+    val result: Set[ZeroCopyUTF8String] = new mutable.HashSet[ZeroCopyUTF8String]()
+    val partIterator = partKeyIndex.partIdsFromFilters(filter, startTime, endTime)
+    while(partIterator.hasNext && result.size < limit) {
+      val nextPart = getPartitionFromPartId(partIterator.next())
+      result += ZeroCopyUTF8String(nextPart.partKeyBytes)
+    }
+    result.toList
+  }
+
+  private def getPartitionFromPartId(partId: Int): TimeSeriesPartition = {
+    var nextPart = partitions.get(partId)
+    if (nextPart == UnsafeUtils.ZeroPointer) {
+      partKeyIndex.partKeyFromPartId(partId)
+        .map(partKey => getPartition(partKey.bytes)
```

Comment:
Please use this syntax:  `.map { partKey =>   ... }`

Reply:
this is not applicable in the new code